README.md
ALOOKUP
=======

Simple application that takes a scan of access points in an area, it will then send that list through to google's geolocation service which will then attempt to determine the approximate gps co-ordindates.

Additionally a rudementary caching system has been incorporated that will attempt to cache the results and match furture calls in an effort to reduce the number of requests that are sent to google.

Installation
============

## Manual Method

I suggest that you use virtualenv to setup an isolated python environment to execute and work with this application

```console
git clone https://github.com/jpunwin/alookup.git alookup
virtualenv alookup
alookup/bin/pip install ./alookup
source alookup/bin/activate
cd alookup
pserve production.ini
```

## Docker

```console
docker build -t junwin/alookup https://github.com/jpunwin/alookup.git
docker run -p 8080:8080 junwin/alookup:latest
```

Testing
=======

You can simply use cURL if you want to perform a post request

```console
curl --request POST --data '{
  "apscan_data": [
    {
      "band": "2.4",
      "bssid": "9c:b2:b2:66:c1:be",
      "channel": "5",
      "frequency": 2432,
      "rates": "1.0 - 135.0 Mbps",
      "rssi": -35,
      "security": "wpa-psk",
      "ssid": "HUAWEI-B315-C1BE",
      "timestamp": 1522886457.0,
      "vendor": "HUAWEI TECHNOLOGIES CO.,LTD",
      "width": "20"
    },
    {
      "band": "2.4",
      "bssid": "84:78:ac:b9:76:19",
      "channel": "1",
      "frequency": 2412,
      "rates": "6.5 - 270.0 Mbps",
      "rssi": -56,
      "security": "wpa-eap",
      "ssid": "1 Telkom Connect",
      "timestamp": 1522886457.0,
      "vendor": "Cisco Systems, Inc",
      "width": "20"
    },
    {
      "band": "2.4",
      "bssid": "c0:a0:bb:c4:10:d6",
      "channel": "1",
      "frequency": 2412,
      "rates": "1.0 - 54.0 Mbps",
      "rssi": -66,
      "security": "wpa-psk",
      "ssid": "default",
      "timestamp": 1522886457.0,
      "vendor": "D-Link International",
      "width": "40"
    },
    {
      "band": "5.0",
      "bssid": "84:78:ac:b9:76:16",
      "channel": "56",
      "frequency": 5280,
      "rates": "6.5 - 270.0 Mbps",
      "rssi": -82,
      "security": "wpa-eap",
      "ssid": "1 Telkom Connect",
      "timestamp": 1522886457.0,
      "vendor": "Cisco Systems, Inc",
      "width": "20"
    },
    {
      "band": "5.0",
      "bssid": "e8:1d:a8:28:a6:6c",
      "channel": "36",
      "frequency": 5180,
      "rates": "6.0 - 866.7 Mbps",
      "rssi": -65,
      "security": "open",
      "ssid": "@VAST",
      "timestamp": 1522886457.0,
      "vendor": "Ruckus Wireless",
      "width": "80"
    },
    {
      "band": "2.4",
      "bssid": "e8:1d:a8:28:a6:68",
      "channel": "8",
      "frequency": 2447,
      "rates": "6.0 - 135.0 Mbps",
      "rssi": -52,
      "security": "open",
      "ssid": "@VAST",
      "timestamp": 1522886457.0,
      "vendor": "Ruckus Wireless",
      "width": "20"
    },
    {
      "band": "2.4",
      "bssid": "e8:1d:a8:68:a6:68",
      "channel": "8",
      "frequency": 2447,
      "rates": "6.0 - 135.0 Mbps",
      "rssi": -53,
      "security": "open",
      "ssid": "McDonalds@VAST",
      "timestamp": 1522886457.0,
      "vendor": "Ruckus Wireless",
      "width": "20"
    },
    {
      "band": "2.4",
      "bssid": "84:78:ac:b9:76:1a",
      "channel": "1",
      "frequency": 2412,
      "rates": "6.5 - 270.0 Mbps",
      "rssi": -57,
      "security": "open",
      "ssid": "1 Telkom Guest",
      "timestamp": 1522886457.0,
      "vendor": "Cisco Systems, Inc",
      "width": "20"
    },
    {
      "band": "5.0",
      "bssid": "84:78:ac:b9:76:15",
      "channel": "56",
      "frequency": 5280,
      "rates": "6.5 - 270.0 Mbps",
      "rssi": -82,
      "security": "open",
      "ssid": "1 Telkom Guest",
      "timestamp": 1522886457.0,
      "vendor": "Cisco Systems, Inc",
      "width": "20"
    },
    {
      "band": "5.0",
      "bssid": "e8:1d:a8:68:a6:6c",
      "channel": "36",
      "frequency": 5180,
      "rates": "6.0 - 866.7 Mbps",
      "rssi": -66,
      "security": "open",
      "ssid": "McDonalds@VAST",
      "timestamp": 1522886457.0,
      "vendor": "Ruckus Wireless",
      "width": "80"
    }
  ]
}' http://localhost:8080
```

TODO
====

Some things that need to be done
 - Need to add in the RedisBackend for the caching system
 - Add kubernetes documentation
 - Could theoretically be easily converted to Python3
 - Add some direct testing of the caching system
 - Cache entries need a TTL, the DictBackend will need to invalidate cached items after a configured amount of time


CHANGES.md
0.1
---

- Initial Version
 - Restful view
 - Google geolocation interface
- Caching
 - Added DictBackend
 - Decorator
- Dockerised
 - Added Entry Script
 - Setup the Dockerfile
- Documentation
 - Added Readme.md
 - Documented inline code





tests/conftest.py
import os
from pyramid.paster import get_appsettings
from pyramid.scripting import prepare
from pyramid.testing import DummyRequest
import pytest
import webtest

from alookup import main


def pytest_addoption(parser):
    parser.addoption('--ini', action='store', metavar='INI_FILE')

@pytest.fixture(scope='session')
def ini_file(request):
    # potentially grab this path from a pytest option
    return os.path.abspath(request.config.option.ini or 'testing.ini')

@pytest.fixture(scope='session')
def app_settings(ini_file):
    return get_appsettings(ini_file)

@pytest.fixture(scope='session')
def app(app_settings):
    return main({}, **app_settings)

@pytest.fixture
def testapp(app):
    testapp = webtest.TestApp(app, extra_environ={
        'HTTP_HOST': 'example.com',
    })

    return testapp

@pytest.fixture
def app_request(app):
    """
    A real request.

    This request is almost identical to a real request but it has some
    drawbacks in tests as it's harder to mock data and is heavier.

    """
    env = prepare(registry=app.registry)
    request = env['request']
    request.host = 'example.com'

    yield request
    env['closer']()

@pytest.fixture
def dummy_request(app):
    """
    A lightweight dummy request.

    This request is ultra-lightweight and should be used only when the
    request itself is not a large focus in the call-stack.

    It is way easier to mock and control side-effects using this object.

    - It does not have request extensions applied.
    - Threadlocals are not properly pushed.

    """
    request = DummyRequest()
    request.registry = app.registry
    request.host = 'example.com'

    return request

tests/test_views.py
"""Summary

Attributes:
    api_key (str): Description
    test_json_body (TYPE): Description
"""
from alookup.views.default import lookup_view
from alookup.lib.caching import setup_cache_fromsettings
from alookup.lib.caching import GoogleGeocodeCache
from pyramid.testing import DummyRequest
import pytest

test_json_body = {
    "apscan_data": [
      {
        "band": "2.4",
        "bssid": "9c:b2:b2:66:c1:be",
        "channel": "5",
        "frequency": 2432,
        "rates": "1.0 - 135.0 Mbps",
        "rssi": -35,
        "security": "wpa-psk",
        "ssid": "HUAWEI-B315-C1BE",
        "timestamp": 1522886457.0,
        "vendor": "HUAWEI TECHNOLOGIES CO.,LTD",
        "width": "20"
      },
      {
        "band": "2.4",
        "bssid": "84:78:ac:b9:76:19",
        "channel": "1",
        "frequency": 2412,
        "rates": "6.5 - 270.0 Mbps",
        "rssi": -56,
        "security": "wpa-eap",
        "ssid": "1 Telkom Connect",
        "timestamp": 1522886457.0,
        "vendor": "Cisco Systems, Inc",
        "width": "20"
      },
      {
        "band": "2.4",
        "bssid": "c0:a0:bb:c4:10:d6",
        "channel": "1",
        "frequency": 2412,
        "rates": "1.0 - 54.0 Mbps",
        "rssi": -66,
        "security": "wpa-psk",
        "ssid": "default",
        "timestamp": 1522886457.0,
        "vendor": "D-Link International",
        "width": "40"
      },
      {
        "band": "5.0",
        "bssid": "84:78:ac:b9:76:16",
        "channel": "56",
        "frequency": 5280,
        "rates": "6.5 - 270.0 Mbps",
        "rssi": -82,
        "security": "wpa-eap",
        "ssid": "1 Telkom Connect",
        "timestamp": 1522886457.0,
        "vendor": "Cisco Systems, Inc",
        "width": "20"
      },
      {
        "band": "5.0",
        "bssid": "e8:1d:a8:28:a6:6c",
        "channel": "36",
        "frequency": 5180,
        "rates": "6.0 - 866.7 Mbps",
        "rssi": -65,
        "security": "open",
        "ssid": "@VAST",
        "timestamp": 1522886457.0,
        "vendor": "Ruckus Wireless",
        "width": "80"
      },
      {
        "band": "2.4",
        "bssid": "e8:1d:a8:28:a6:68",
        "channel": "8",
        "frequency": 2447,
        "rates": "6.0 - 135.0 Mbps",
        "rssi": -52,
        "security": "open",
        "ssid": "@VAST",
        "timestamp": 1522886457.0,
        "vendor": "Ruckus Wireless",
        "width": "20"
      },
      {
        "band": "2.4",
        "bssid": "e8:1d:a8:68:a6:68",
        "channel": "8",
        "frequency": 2447,
        "rates": "6.0 - 135.0 Mbps",
        "rssi": -53,
        "security": "open",
        "ssid": "McDonalds@VAST",
        "timestamp": 1522886457.0,
        "vendor": "Ruckus Wireless",
        "width": "20"
      },
      {
        "band": "2.4",
        "bssid": "84:78:ac:b9:76:1a",
        "channel": "1",
        "frequency": 2412,
        "rates": "6.5 - 270.0 Mbps",
        "rssi": -57,
        "security": "open",
        "ssid": "1 Telkom Guest",
        "timestamp": 1522886457.0,
        "vendor": "Cisco Systems, Inc",
        "width": "20"
      },
      {
        "band": "5.0",
        "bssid": "84:78:ac:b9:76:15",
        "channel": "56",
        "frequency": 5280,
        "rates": "6.5 - 270.0 Mbps",
        "rssi": -82,
        "security": "open",
        "ssid": "1 Telkom Guest",
        "timestamp": 1522886457.0,
        "vendor": "Cisco Systems, Inc",
        "width": "20"
      },
      {
        "band": "5.0",
        "bssid": "e8:1d:a8:68:a6:6c",
        "channel": "36",
        "frequency": 5180,
        "rates": "6.0 - 866.7 Mbps",
        "rssi": -66,
        "security": "open",
        "ssid": "McDonalds@VAST",
        "timestamp": 1522886457.0,
        "vendor": "Ruckus Wireless",
        "width": "80"
      }
    ]
}

api_key = "AIzaSyCCK6hPzvUI1_XbDCV4pC1HN_6bneUejYc"

@pytest.fixture
def apscan_data(app):
    """Summary
    
    Args:
        app (TYPE): Description
    
    Returns:
        TYPE: Description
    """
    app.registry.settings["geolocate.api_key"] = api_key
    app.registry.settings["cache.enabled"] = "True"
    setup_cache_fromsettings(app.registry.settings)
    request = DummyRequest(json_body=test_json_body, method="POST")
    request.registry = app.registry
    request.host = 'example.com'
    data = test_json_body.copy()
    del data["apscan_data"][5]
    request_partial = DummyRequest(json_body=data, method="POST")
    request_partial.registry = app.registry
    request_partial.host = 'example.com'
    return request, request_partial

def test_perform_lookup(apscan_data):
    """Summary
    
    Args:
        apscan_data (TYPE): Description
    """
    data_full, data_partial = apscan_data
    res = lookup_view(data_full)
    assert data_full.response.status_int == 200
    assert res != None
    assert len(GoogleGeocodeCache.backend.data.keys()) > 0
    assert res["location"] != None
    res = lookup_view(data_partial)
    assert data_partial.response.status_int == 200
    assert res != None
    assert len(GoogleGeocodeCache.backend.data.keys()) > 0
    assert res["location"] != None


alookup/lib/glookup.py
"""
Google Lookup System
"""
import logging
import requests
import json
from caching import GoogleGeocodeCache as cache

log = logging.getLogger("google.geolocation")


class InvalidConfiguration(Exception):

    """InvalidConfiguration exception is thrown on an instance of a required configuration not being set
    """

    pass


class InvalidAPRequest(Exception):

    """InvalidAPRequest exception is thrown when the json request body is missing required information, i.e. is malformed
    """

    pass


def format_ap_entry(ap_entry):
    """Processes the Access Point dictionary and produce a stripped down entry for use in the request dataset

    Args:
        ap_entry (DICT): Access Point dict entry
    """
    return dict(
        macAddress=ap_entry["bssid"],
        signalStrength=int(ap_entry["rssi"]),
        age=0,
        channel=int(ap_entry["channel"])
    )


def normalise_request(apscan):
    """Attempts to locate the access point information and return it as a list of access points

    Args:
        apscan (DICT|LIST): Description
    """

    out = []
    if type(apscan) == dict:
        # We need to itterate through the key values and
        # look for a list of dictionaries where a bssid key exists
        for k, v in apscan.items():
            if type(v) == list:
                out = normalise_request(v)
                if out:
                    return out
    elif type(apscan) == list:
        for ent in apscan:
            formatted_ent = format_ap_entry(ent)
            if formatted_ent:
                out += [formatted_ent]
    return out


@cache
def geolocate(apscan, settings={}):
    """gelocate takes a list of dictionaries items that are sent through to google via a post request the resulting is then parsed and returned

    Args:
        apscan (LIST): Scan of Wifi Access points that should be submitted to google geolocation interface
        settings (dict, optional): Pyramid Settings instance

    Returns:
        TYPE: Description

    Raises:
        InvalidConfiguration: Description
    """
    log.debug("Request for: %s" % apscan)
    api_key = settings.get("geolocate.api_key")
    if not api_key:
        raise InvalidConfiguration(
            "geolocate.api_key has not been set in the config")
    url = "https://www.googleapis.com/geolocation/v1/geolocate?key=" + api_key
    _post_info = dict(wifiAccessPoints=apscan)
    response = requests.post(url, data=json.dumps(_post_info))
    if response.status_code == 200:
        return response.json()


def perform_lookup(apscan, settings=None):
    """Sends a request to google's geolocation service to lookup the location
    based on the Access Point list provided in the request

    Args:
        apscan (DICT|LIST): a list of access points
    """
    # We need to start by assuming we may not get the information
    # In the correct format
    if not settings:
        settings = get_current_registry().settings
    apscan = normalise_request(apscan)
    if not apscan:
        raise InvalidAPRequest("Invalid Request")
    return geolocate(apscan, settings)


alookup/views/default.py
from pyramid.view import view_config
from pyramid.httpexceptions import HTTPBadRequest
from pyramid.httpexceptions import HTTPInternalServerError
from ..lib.glookup import perform_lookup
from ..lib.glookup import InvalidAPRequest
from ..lib.glookup import InvalidConfiguration


@view_config(route_name='root', renderer='json')
def lookup_view(request):
    """Summary

    Args:
        request (pyramid.request): ...

    Returns:
        LIST: Description

    Raises:
        HTTPBadRequest: On an error in the request, an HTTPBadRequest is raised
    """
    request_payload = None
    try:
        request_payload = request.json_body
    except ValueError, exc:
        raise HTTPBadRequest(str(exc))
    try:
        return perform_lookup(request_payload, request.registry.settings)
    except InvalidAPRequest, exc:
        raise HTTPBadRequest(str(exc))
    except InvalidConfiguration, exc:
        raise HTTPInternalServerError(str(exc))


def includeme(config):
    config.add_route('root', '/')


alookup/lib/caching.py
"""Caching for google wifi access point geolocation requests

Example:
    from alookup.lib.caching import GoogleGeocodeCache as cache

    ...

    @cache
    def function_that_needs_caching(apscan, ...):
        #Do your stuff
        ...
        return output

    #Don't forget to set it up when you init caching using config.include
    def main(global_config, **settings):
        with Configurator(settings=settings) as config:
            #Set up Caching
            config.include('.lib.caching')

This module implements
 - Backend Support
  - DictBackend [in memory]
  - RedisBackend *TODO
 - Indexed Access Point approximation for fast caching

Attributes:
    backend_urn_re (re): Regular expression for matching the backends
    log (logging): Logger instance for logging for this module

Todo:
 - RedisBackend needs to be implemented
 - TTL Cache invalidation needs to be created
"""
from functools import update_wrapper
from pyramid.threadlocal import get_current_registry
from pyramid.settings import asbool
from math import ceil
import uuid
import re
import logging

log = logging.getLogger("alookup.caching")


class DictBackend(object):

    """Implements an in memory dictionary backend for use in caching
    """
    
    def __init__(self, cnf=None):
        """Initialize the data memory storage
        """
        log.info("Caching started with DictBackend")
        self.data = {}

    def flush(self):
        """Empty the storage backend of all it's cached entries
        """
        self.data = {}

    def setkey(self, key, value):
        """Set the value for a specific key
        
        Args:
            key (str): Key to update the value
            value (*): Value to set
        """
        self.data[key] = value

    def appendkey(self, key, value):
        """Addes the value to the value at the key

        This assumes the value is part of a list so it will always append it to the list, 
        if the key does not exist it will create the list with the value as the only item in
        that list
        
        Args:
            key (str: Key for the list
            value (*): Value to add to the list
        
        Raises:
            TypeError: If the key exists in the backend but the value is not a list it will raise a TypeError
        """
        if key in self.data:
            if type(self.data[key]) != list:
                raise TypeError("Data type is not a list")
            else:
                self.data[key].append(value)
        else:
            self.data[key] = [value]

    def getkey(self, key):
        """Get the value at a specific key
        
        Args:
            key (str): Key to get the value
        
        Returns:
            *: Value at the key
        """
        return self.data.get(key)

    def iskey(self, key):
        """Tests if the key exists in the backend
        
        Args:
            key (str): Key to test
        
        Returns:
            boolean: If the key exists then we return True
        """
        return key in self.data

class RedisBackend(object):
    """Implements the RedisBackend

    On a distributed production environment it is useful to have a transient caching system

    i.e. Multiple separate instances of the application can be run and 
    share the same cache backend across all of them, this means that on high volume
    systems you get the caching benefit of all of these

    TODO:
     - Implement this backend

    """
    
    def __init__(self, cnf):
        raise NotImplemented("Definately can be done if you wanted to scale the system")

backend_urn_re = re.compile("^(dict|redis\:\/\/(.*){3,5})$")

_backends = {
    "dict": DictBackend,
    "redis": RedisBackend
}

class GoogleGeocodeCache(object):

    """Cache Decorator is used to cleanly apply the caching
    
    Attributes:
        backend (backend): Instance of the active backend
        depth (int): How many access point cache entries will we search through before we give up
        func (func): Decorated function
        is_enabled (bool): Is caching enabled
        match_required (int): Percentage of the match that is required to consider them the same
        signal_bands (dict): Dictionary lookup to group signals into sets (bands)
    """
    
    is_enabled = False

    backend = None

    depth = 6

    match_required = 80

    signal_bands = {
        40: "A",
        50: "B",
        60: "C",
        70: "D",
        80: "E",
        9999: "F"
    }

    def __init__(self, func):
        """When we decorate a function it inits the class which replaces the decorated function
        
        Args:
            func (func): Decorated function
        """
        update_wrapper(self, func)
        self.func = func

    def generate_cache_key(self, ent):
        """Takes an apscan entry and determines a backend key

        <SSID:MACADDRESS>::<Signal Band>
        
        Args:
            ent (dict): Apscan entry
        
        Returns:
            str: Key derived from the access point information
        """
        k = [i for i in sorted(self.signal_bands.keys()) if i >= -ent["signalStrength"]][0]
        sband = self.signal_bands[k]
        return "%s::%s" % (ent["macAddress"], sband)

    def save(self, apscan, result):
        """Cache the result
        
        Args:
            apscan (list): List of access point scan entries
            result (dict): The result of executing the decorated self.func
        """
        apscan = sorted(apscan, key=lambda x:x["signalStrength"], reverse=True)
        _id = str(uuid.uuid4())
        _keylist = []
        for ent in apscan:
            k = self.generate_cache_key(ent)
            self.backend.appendkey(k, _id)
            _keylist.append(k)
        _data = {
            "request": apscan,
            "result": result,
            "keylist": _keylist
        }
        self.backend.setkey(_id, _data)

    def lookup(self, apscan):
        """Attempts to lookup the apscan entries to see if a cached entry exists
        
        Args:
            apscan (list): List of access points scan entries
        
        Returns:
            dict(nullable): If we can find the cache entry we return it otherwise null is returned
        """
        apscan = sorted(apscan, key=lambda x:x["signalStrength"], reverse=True)
        _ent_cnt = len(apscan)
        _uuid_list = []

        _keys = [self.generate_cache_key(ent)for ent in apscan]
        for k in _keys:
            _uuid_res = self.backend.getkey(k)
            if _uuid_res:
                _uuid_list += _uuid_res
        uuid_cnts = dict((i,1) for i in _uuid_list)
        for key in uuid_cnts:
            uuid_cnts[key] = _uuid_list.count(key)
        uuid_list = sorted(uuid_cnts.keys(), key=lambda x:uuid_cnts[x], reverse=True)
        if uuid_list:
            for u_key in uuid_list[0:self.depth]:
                res = self.backend.getkey(u_key)
                diff_cnt = len(set(_keys).symmetric_difference(set(res["keylist"])))
                if ceil((float(diff_cnt) / float(_ent_cnt)) * 100) < 100-self.match_required:
                    return res["result"]

    def __call__(self, *a):
        """Call wrapper that proxies the function call
        
        Args:
            *a: Argument list
        
        Returns:
            *: Result of executing the function
        """
        if self.is_enabled:
            res = self.lookup(a[0])
            if res:
                log.info("Cached result for %s" % a[0])
                return res
            else:
                res = self.func(*a)
                self.save(a[0], res)
                log.info("Added caching for %s" % a[0])
                return res
        else:
            return self.func(*a)


def setup_cache_fromsettings(settings):
    """Sets up the caching from the provided settings
    
    Args:
        settings (DICT): Pyramid Settings
    """
    GoogleGeocodeCache.is_enabled = asbool(settings.get("cache.enabled"))
    GoogleGeocodeCache.depth = int(settings.get("cache.depth", 6))
    GoogleGeocodeCache.match_required = int(settings.get("cache.match_required", 80))
    if backend_urn_re.search(settings.get("cache.backend", "dict")):
        if "://" in settings.get("cache.backend", ""):
            backend, cnf = settings.get("cache.backend").split("://")
        else:
            backend = settings.get("cache.backend", "dict")
            cnf = ""
        if backend in _backends:
            _resolved_backend = _backends[backend]
            GoogleGeocodeCache.backend = _resolved_backend(cnf)
    
def includeme(config):
    setup_cache_fromsettings(config.registry.settings)